src/coord/src/coord.rs

```diff
@@ -5274,3 +5198,94 @@ impl Coordinator {
+    /// Specifically, all read timestamps will be greater or equal to all previously reported write timestamps,
+    /// and strictly less than all subsequently emitted write timestamps.
+    pub struct Timeline {
+        state: TimelineState,
+        advance_to: Option<mz_repr::Timestamp>,
+    }
+
+    impl Timeline {
+        /// Create a new timeline, starting at the indicated time.
+        pub fn new(initially: mz_repr::Timestamp) -> Self {
+            Self {
+                state: TimelineState::Writing(initially),
+                advance_to: Some(initially),
+            }
+        }
+
+        /// Acquire a new timestamp for writing.
+        ///
+        /// This timestamp will be strictly greater than all prior values of `self.read_ts()`,
+        /// and less than or equal to all subsequent values of `self.read_ts()`.
+        pub fn write_ts(&mut self) -> mz_repr::Timestamp {
+            match self.state {
+                TimelineState::Writing(ts) => ts,
+                TimelineState::Reading(ts) => {
+                    self.state = TimelineState::Writing(ts + 1);
+                    ts + 1
+                }
+            }
+        }
+        /// Acquire a new timestamp for reading.
+        ///
+        /// This timestamp will be greater or equal to all prior values of `self.write_ts()`,
+        /// and strictly less than all subsequent values of `self.write_ts()`.
+        pub fn read_ts(&mut self) -> mz_repr::Timestamp {
+            match self.state {
+                TimelineState::Reading(ts) => ts,
+                TimelineState::Writing(ts) => {
+                    self.state = TimelineState::Reading(ts);
+                    self.advance_to = Some(ts + 1);
+                    ts
+                }
+            }
+        }
+        /// Electively advance the tracked times.
+        ///
+        /// If `lower_bound` is strictly greater than the current time (of either state), the
+        /// resulting state will be `Writing(lower_bound)`.
+        pub fn fast_forward(&mut self, lower_bound: mz_repr::Timestamp) {
+            match self.state {
+                TimelineState::Writing(ts) => {
+                    if lower_bound > ts {
+                        self.advance_to = Some(lower_bound);
+                        self.state = TimelineState::Writing(lower_bound);
+                    }
+                }
+                TimelineState::Reading(ts) => {
+                    if lower_bound > ts {
+                        // This may result in repetition in the case `lower_bound == ts + 1`.
+                        // This is documented as fine, and concerned users can protect themselves.
+                        self.advance_to = Some(lower_bound);
+                        self.state = TimelineState::Writing(lower_bound);
+                    }
+                }
+            }
+        }
+        /// Whether and to what the next value of `self.write_ts() has advanced since this method was last called.
```

Comment:
If my understanding above is correct, I think a small tweak to this docstring makes it all more clear! 

```suggestion
        /// Whether and to what the *next* call to `self.write_ts()` will advance to since this method was last called.
```

Reply:
Ugh, idk, my wording is still confusing. I'm sure there is some clear way to explain it, though!